src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,143 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final String ERROR_MESSAGE_OUT_OF_BOUNDS = "Index out of bound: ";
+    private Node<T> head;
+    private Node<T> tail;
+
+    private class Node<T> {
+        private Node<T> next;
+        private Node<T> prev;
+        private T value;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            tail.next = new Node<>(tail, value, null);
+            tail = tail.next;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size() || index < 0) {
+            throw new IndexOutOfBoundsException(ERROR_MESSAGE_OUT_OF_BOUNDS + index);
+        }
+        if (head == null) {
+            head = new Node<T>(null, value, null);
+            tail = head;
+        } else if (size() == index) {
+            tail.next = new Node<>(tail, value, null);
+            tail = tail.next;
+        } else {
+            Node<T> nodeByIndex = findNodeByIndex(index);
+            Node<T> newNode = new Node<>(nodeByIndex.prev, value, nodeByIndex);
+            newNode.next.prev = newNode;
+            if (index != 0) {
+                newNode.prev.next = newNode;
+            } else {
+                head = newNode;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            tail.next = new Node<>(tail, element, null);
+            tail = tail.next;
+        }
```

Comment:
what if our linked list is empty, and we want to addAll several elements?
let's call `add(value)` method for each element